WPMedia/fusion-news-theme-blocks
# Lazy load - Top Table List and Small Promo

## Description

Implemented the Engine SDK LazyLoad component into the Top Table List and Small Promo block along with server side checks to ensure when the custom field lazyLoad is enabled the feature is not rendered server side. Moving the rendering and data fetching over to the client side.


## Jira Ticket
- [TMEDIA-116](https://arcpublishing.atlassian.net/browse/TMEDIA-116)

## Test Steps

1. Checkout this branch `git checkout lazy-load-top-table-list-small-promo`
2. Run fusion repo with linked blocks `npx fusion start -f -l @wpmedia/top-table-list-block,@wpmedia/small-promo-block`
3. Either on a new page or existing page toggle on the new custom field option "lazyLoad" for a top table list and small promo feature that appears at more than 300px below the viewport.
4. View the published page
5. Verify the HTML for each feature is not present on page load - You can disable JavaScript in dev tools and verify the features you enabled lazyLoad for do not render on server render
6. Verify when you scroll the page that as you begin to get nearer the features there are network calls to fetch the content needed for each feature you have enabled lazyLoad for.


## Effect Of Changes

New Custom field
<img width="289" alt="LazyLoad - Custom field option" src="https://user-images.githubusercontent.com/868127/110464506-1ee56a80-80cb-11eb-93e4-f1f76ffb636e.png">

Showing page weight saving by have lazyLoad enabled on initial server side render
Example shows top page with 4 top table list features, with 2 having lazy load enabled, bottom is same top table list features with none using lazyLoad

![lazyLoad - Page load - top table list - with and without](https://user-images.githubusercontent.com/868127/110464632-450b0a80-80cb-11eb-95d9-f43d1398f020.png)

Video showing content being fetched client side as you scroll the page

https://user-images.githubusercontent.com/868127/110464664-50f6cc80-80cb-11eb-9f3a-8718a3233ff6.mp4



## Author Checklist
_The author of the PR should fill out the following sections to ensure this PR is ready for review._
please explain why (so that we can fix it whenever it gets refactored).

## Reviewer Checklist 
_The reviewer of the PR should copy-paste this template into the review comments on review._
